platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/builder/PlatformBuilder.java

```diff
@@ -166,15 +168,18 @@ public final class PlatformBuilder {
      * @param selfId              the ID of this node
      * @param softwareVersion     the software version of the application
      * @param genesisStateBuilder a supplier that will be called to create the genesis state, if necessary
+     * @param snapshotStateReader a function to read an existing state snapshot, if exists
      */
     @NonNull
     public static PlatformBuilder create(
             @NonNull final String appName,
             @NonNull final String swirldName,
             @NonNull final SoftwareVersion softwareVersion,
-            @NonNull final Supplier<SwirldState> genesisStateBuilder,
+            @NonNull final Supplier<MerkleRoot> genesisStateBuilder,
+            @NonNull final CheckedBiFunction<MerkleDataInputStream, Path, MerkleRoot, IOException> snapshotStateReader,
```

Comment:
I added a comment to this method clarifying the flow. First, a snapshot on disk (if exists) is loaded using the provided loader function. If it throws an exception, a new genesis state is created.

Reply:
Yes, this is exactly inversion of control. Applications may have different merkle state trees. For example, apps like PTT will use `State` as the merkle root, while Hedera will use `MerkleHederaState`. The two params to PlatformBuilder are exactly to enable apps to plug in their merkle tree implementation.

Reply:
I also added a note about `genesisStateBuilder` / `snapshotStateReader` to not return nulls, but rather throw an exception. If a null is returned by `genesisStateBuilder` today, it will result in an NPE somewhere down the stack, and `snapshotStateReader` is no different. I don't see a value in handling nulls explicitly.

Reply:
Merging the two args into one should be possible, but I wouldn't make it as a part of this PR. It will require much more changes: how and when snapshots are done, where snapshot folders are located, and so on. Most of it is currently done by the platform, but it all should be delegated to the app.

Reply:
agree in general with @cody-littley but we should create an issue for that and do it in a follow up.

Reply:
Feel free to mark this comment as resolved (from my perspective). Leaving it open for now in case there is further discussion needed.